cdap-app-fabric/src/main/java/co/cask/cdap/templates/AdapterSpecification.java

```diff
@@ -134,11 +153,16 @@ public static Builder builder(String name, String template) {
     private Resources resources;
     private JsonElement config;
 
-    public Builder(String name, String template) {
+    public Builder(String name, String template, ProgramType programType, String programName) {
       Preconditions.checkArgument(name != null, "Adapter name must be specified.");
       Preconditions.checkArgument(template != null, "Adapter template must be specified.");
+      Preconditions.checkArgument(ADAPTER_PROGRAM_TYPES.contains(programType), "Adapter program type must be on these: "
```

Comment:
nit: *one* of these. 

Also, ``Preconditions`` supports String.format style options. You can use that instead of concatenation